### ISSUE TITLE
How to get baseline result

### ISSUE THREAD

**Issue by L1aoXingyu** [Yu-Wu/One-Example-Person-ReID#3]
Hi, Thanks for your awesome work. I read the paper and found baseline in market1501 is rank1=26.0, mAP=9.0. In my opinion, this is trained using 751 images with 751 ids. But I just try, and I cannot get the baseline result. My rank1=8.6, mAP=3.0. I want to know how to reproduce the baseline result using this code.

**Comment by Yu-Wu** [comment]
@L1aoXingyu 

Thanks for your interest on our paper.

Yes, for the one-example setting, the model is firstly trained using 751 images with 751 ids on Market1501.

To reproduce the baseline results, you can just run this code by "sh run.sh" and observe the first evaluation results. 
It firstly trains a baseline model and then begin to iteratively update the model.
Therefore, after the first 70 training epochs, you can see the evaluation results for the first baseline model.

**Comment by L1aoXingyu** [comment]
I checked code and found that use_selcted_data is True when training at first step_size 
![image](https://user-images.githubusercontent.com/19245504/52764962-1beb2d80-305d-11e9-90dc-98f8bd7c0c35.png)

I think this will use exclusive loss for unlabel data and it's not the baseline result. I just pick 751 images randomly, one image for one person and train 70 epochs, but I can only get rank1=11.4, mAP=3.8. Or I made some mistakes?

**Comment by Yu-Wu** [comment]
@L1aoXingyu 

Thanks for your reply. 

The baseline results are achieved by set "use_selcted_data=False". You can try my code with "use_selcted_data=False", and check the performances of the first trained model. 

This is the baseline and should be consistent with "rank1=26.0, mAP=9.0" in the paper.

As for the results of your own implementation, I think the main reason is the hyper parameter in the optimizer. 
Could you please check the momentum value of the SGD optimizer in your code? You will achieve the same performance with mine if you set the momentum to 0.5, instead of 0.9. 
(For the one-example setting, the model cannot get converged rapidly if you set a very large value for the momentum weight.)

**Comment by L1aoXingyu** [comment]
Many Thanks for your reply without delay.

I just used Adam instead of SGD. As your opinion, the model shouldn't converge rapidly because it will lead to overfitting? I will try SGD and change the momentum to 0.5. Thanks again.

**Comment by Yu-Wu** [comment]
@L1aoXingyu 

I never tried the Adam optimizer. But I think the large momentum for SGD will harm the training. 

Due to the very limited data samples (751 images for 751 classes), the optimization directions of mini-batchs might diverge obviously. Setting momentum to 0.9 indicates 90% of optimization direction comes from the history, instead of current gradient direction.

You may check the training accuracy (i.e., the classification accuracy on the training samples). The training accuracy is always low if you set momentum to 0.9. This is not the "overfitting" issue. Instead, this is "underfitting".

**Comment by L1aoXingyu** [comment]
Actually, my model's classification accuracy is very high after several epochs.
![image](https://user-images.githubusercontent.com/19245504/52769736-23fe9980-306c-11e9-910e-76bc50c60b18.png)

My model is a little bit different from you. I remove last conv's stride, and just add a BatchNorm1d and classification layer.

I just tried your model architecture and check what is the problem.

**Comment by Yu-Wu** [comment]
@L1aoXingyu 

I am not clear about your model architecture. But just now I tried my code with SGD (momentum=0.9), and it achieves "rank-1 9.8%, map 3.3%", which is close to your reported performance. The comparison proves the momentum value is of virtual importance in the One-Example optimization.

Hope this helps.

**Comment by L1aoXingyu** [comment]
Many thanks for your patient answers. I will do some experiments and tell you the results.

**Comment by L1aoXingyu** [comment]
@Yu-Wu 
I changed all settings just like yours, and I can get your baseline result.
![image](https://user-images.githubusercontent.com/19245504/52776378-65974080-307c-11e9-986e-5daba2d80d20.png)
I think this is very tricky. I changed the model architecture, optimzier learning rate, feature extractor learning rate. But I can only get about 14.5% rank1. At last, I change the image preprocess. In my preprocess, I do resize, 10 padding, random crop, random horizontal flip, random erasing. In your preprocess, you just do randomsized crop, flip. 
Actually I don't know why there is such a big drop just because of different preprocess, I will do some experiments to explore the reason.

BTW, you can see my architecture in here https://github.com/L1aoXingyu/reid_baseline. This baseline can outperformance many papers results.

**Comment by L1aoXingyu** [comment]
I do some ablation studies and get the results below

|Setting | Rank1 | mAP 
|-- | -- | -- 
|paper report | 26.0 | 9.0 
|Reproduce | 25.1 | 9.2
|+ RandomErasing | 19.3 (-5.8) | 7.4 (-1.8)
|+ Pad and RandomCrop | 21.1 (-4.0) | 7.5 (-1.7)
|+ Test Euclidean distance | 23.0 (-2.1) | 7.9 (-1.3)
|+ Same learning rate 0.1 (resnet + extra) | 9.6 (-15.5) | 3.5 (-5.5)
|+ Same learning rate 0.01 (resnet + extra) | 21.9 (-3.2) | 8.1 (-1.1)
|+ momentum=0.9 | 2.7 (-22.4) | 1.2 (-7.8)
|+ Adam lr: (resnet: 3.5e-4, extra: 3.5e-5) | 21.9 (-3.2) | 7.7 (-1.5)
|- dropout | 23.5 (-1.6) | 8.6 (-0.6)
|- embedding | 22.4 (-2.7) | 8.3 (-0.9)
|My archi | 19.1 (-6.0) | 6.1 (-3.1)
|My archi + dropout | 27.0 (+1.9) | 9.9 (+0.7)

I found that random erasing will harm the model's performance. Besides, feature extractor parameters with 0.1 times smaller than extra parameters learning rate are very important.

As you say, momentum also plays an important role.

I add dropout in my simple architecture and can get a better result than my reproduction. I think one-shot is very hard to train especially at the first few steps with a large learning rate or momentum. It will push the model to a local minimum or overfitting. Actually, I can see that classification accuracy is almost 100% in the one-shot training set.

**Comment by Yu-Wu** [comment]
@L1aoXingyu 

Many thanks for your valuable experiments results! These results are very helpful in the one-example research.

For one-example image-based re-ID, the image data are very limited, and thus the optimization might be tricky. It is not a surprise that the commonly used data augmentation methods fail since the training is very sensitive to the limited input data. By the way, I think it might need more training iterations for the model to memorize if we apply these data augmentation methods.

One needs to be very careful to train the model under the one-example situation or small datasets. I believe your experimental results provide very useful clues for the few examples learning. In addition, your baseline repo is very helpful and the performance is really promising.

Thanks for your reporting again!

**Comment by Yu-Wu** [comment]
1.
